src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -19,7 +23,54 @@ internal static class DeserializationHelpers
         internal static void ApplyProperty(ODataProperty property, IEdmStructuredTypeReference resourceType, object resource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
         {
-            IEdmProperty edmProperty = resourceType.FindProperty(property.Name);
+            WebApiRequestMessage request = null;
+            if(readContext != null)
+            {
+                request = (WebApiRequestMessage)readContext.InternalRequest;
+            }
+
+            bool enableCaseInsensitiveModelBinding = false;
+
+            if(request != null)
+            {
+#if NETCORE
+                // In AspNetCore, we add our setting by adding a service to the ServiceCollection as below:
+                // services.TryAddEnumerable(ServiceDescriptor.Singleton<IODataModelBinderSettings>(new ODataModelBinderSettings { EnableCaseInsensitiveModelBinding = true }));
+                // We need to resolve the service so that we can get our setting.
+                var settings = request.innerRequest.HttpContext.RequestServices.GetService<IODataModelBinderSettings>();
+                if(settings != null)
+                {
+                    enableCaseInsensitiveModelBinding = settings.EnableCaseInsensitiveModelBinding;
+                }
+#else
+                // In AspNet classic, we add our setting to HttpConfiguration.Properties as below:
+                // configuration.Properties[EnableCaseInsensitiveModelBindingKey] = true;
+                var container = request.innerRequest.GetConfiguration();
+                if(container != null)
+                {
+                    enableCaseInsensitiveModelBinding = container.HasEnabledCaseInsensitiveModelBinding();
+                }
+#endif
+            }
+
+            IEdmProperty edmProperty;
+
+            if (enableCaseInsensitiveModelBinding)
+            {
+                IDictionary<string, string> propertyNamesMapping = new Dictionary<string, string>();
```

Comment:
>IDictionary<string, string> propertyNamesMapping = new Dictionary<string, string>(); [](start = 16, length = 84)

Isn't it a bit expensive to create this each time? Can we create the mapping table on first use and then cache it?

Reply:
#Resolved